pytorch_lightning/callbacks/batch_size_finder.py

```diff
@@ -0,0 +1,423 @@
+            and hasattr(pl_module, "hparams")
+            and self.batch_arg_name in pl_module.hparams
+        ):
+            rank_zero_warn(
+                f"Field `model.{self.batch_arg_name}` and `model.hparams.{self.batch_arg_name}` are mutually exclusive!"
+                f" `model.{self.batch_arg_name}` will be used as the initial batch size for scaling."
+                " If this is not the intended behavior, please remove either one."
+            )
+
+        # TODO: check if this can be enabled (#4040)
+        if not trainer._data_connector._train_dataloader_source.is_module():
+            raise MisconfigurationException(
+                "The batch scaling feature cannot be used with dataloaders passed directly to `.fit()`."
+                " Please disable the feature or incorporate the dataloader into the model."
+            )
+
+        # Arguments we adjust during the batch size finder, save for restoring
+        self._dump_params(trainer)
+
+        # Set to values that are required by the algorithm
+        self._reset_params(trainer)
+
+        # Save initial model, that is loaded after batch size is found
+        save_path = os.path.join(trainer.default_root_dir, f".scale_batch_size_temp_model_{uuid.uuid4()}.ckpt")
+        trainer.save_checkpoint(save_path)
+
+        if trainer.progress_bar_callback:
+            trainer.progress_bar_callback.disable()
+
+        new_size, _ = self._adjust_batch_size(trainer, value=self.init_val)
+
+        if self.mode == "power":
+            new_size = self._run_power_scaling(trainer, pl_module, new_size)
+        elif self.mode == "binsearch":
+            new_size = self._run_binary_scaling(trainer, pl_module, new_size)
+
+        garbage_collection_cuda()
+
+        if trainer.is_global_zero:
+            trainer.checkpoint_connector.restore(save_path)
+            fs = get_filesystem(save_path)
+            if fs.exists(save_path):
+                fs.rm(save_path)
+
+        # global step and current epoch are incremented before saved in checkpoint
+        trainer.fit_loop.global_step -= 1
+        trainer.fit_loop.current_epoch -= 1
+
+        self._restore_params(trainer)
+
+        if trainer.progress_bar_callback:
+            trainer.progress_bar_callback.enable()
+
+        print(f"New batch size: {new_size}")
+        self.optimal_batch_size = new_size
+
+        if self._early_exit:
+            raise _TunerExitException()
+
+    def _run_power_scaling(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule", new_size: int) -> int:
+        """Batch scaling mode where the size is doubled at each iteration until an OOM error is encountered."""
+        for _ in range(self.max_trials):
+            garbage_collection_cuda()
+
+            try:
+                self._try_loop_run(trainer)
+                new_size, changed = self._adjust_batch_size(trainer, factor=2.0, desc="succeeded")
+
+                if changed:
```

Comment:
changed and new_size should be reduced between all ranks. This can be a source of process divergence here.

Reply:
will address this when DDP support will be tested/added. Adding a Misconfiguration now that this will not work with DDP.

Reply:
Curious. What issues are limiting us to add DDP support?